### ISSUE TITLE
Special character in gloss leads to errors

### ISSUE THREAD

**Issue by ocrasborn** [Signbank/NGT-signbank#194]
I get an error when I try to change an E to an Ë in a gloss (BACTERIEN). I click it away, exit edit mode, and then searching for the sign I see in the list view that the change has been applied. But when I try to go to the detail view again, I get another (full-page) error message.
We should be able to use characters like this in glosses. I thought it had come up before, but can't find it in old issues.

**Comment by Woseseltops** [comment]
> I thought it had come up before, but can't find it in old issues.

It has - I think I solved over 20 Unicode problems when we started, but since then new Unicode problems have arrived that we simply never noticed because the idgloss (almost?) never have weird chars.

This issue consisted of two problems:
* Apparently Django does not allow weird chars in search queries; I now search for the current gloss based on the pk.
* The idgloss is part of the url to the video, and urls cannot handle weird chars. I tried three solutions to change weird chars into something different, like BACTERI\&#203\;N-3185.mp4, but all of the solutions resulted in chars that also did not work well in urls, like \ and &. The best solution would be to use percent encoding, but that is some more work, and I wanted to check whether it would be worth the effort. Is it, @ocrasborn ? For now, weird chars are filtered out, I renamed the file manually:

https://signbank.science.ru.nl//static/glossvideo/BA/BACTERIN-3185.mp4

**Comment by henrinie** [comment]
Hi, I too had lots of unicode problems when I started working on FinSL due to letters like ÄÖÅ in the Finnish language.
The video URL's with those letters work for us (or me at least), I am not sure if the url is properly encoded, but it works.
I don't remember how I fixed the unicode errors we had, but there are at least a few differences in our codebases that I can point out:

\__str__ and \__unicode__ in models.py/Gloss
https://github.com/Signbank/NGT-signbank/blob/master/signbank/dictionary/models.py#L226
https://github.com/Signbank/FinSL-signbank/blob/master/signbank/dictionary/models.py#L388

What if you change the \__str__ to \__unicode__, does that work?
Every time you make a string out of idgloss, it uses \__str__ instead of \__unicode__ (at least I think so).
Like in here for example (the idgloss part): https://github.com/Signbank/NGT-signbank/blob/master/signbank/dictionary/models.py#L543

**Comment by henrinie** [comment]
You could also try this:
https://github.com/Signbank/FinSL-signbank/wiki/Install#apache-envvars

**Comment by Woseseltops** [comment]
Thanks for the useful background info, @henrinie !

> What if you change the __str__ to __unicode__, does that work?

I liked the idea so I gave it a try and made it exactly as in the Finnish SB, but unfortunately the generated video url still looked like this: u'/var/www2/signbank/live/repo/signbank/media//glossvideo/BA/BACTERI\xcbN-3185.mp4'

So no :(.

> Every time you make a string out of idgloss, it uses __str__ instead of __unicode__ (at least I think so).
> If I remember correctly, in python3 __str__ is actually __unicode__,

I think so too!

> You could also try this:

Interesting, I didn't know about this! Unfortunately, I won't help for SB, as I don't have the rights to play around with server configs there :/

**Comment by henrinie** [comment]
@Woseseltops what encoding is your database using? Could that be a problem?

**Comment by ocrasborn** [comment]
@Woseseltops : looking ahead at including Chinese Sign Language in the database, including Chinese glosses in a separate field, it'd be good to sort this out properly I think. Also thinking of ELAN and the need to have predictable URLs.

**Comment by henrinie** [comment]
If you are planning to have Chinese glosses, this issue should be solved thoroughly. I don't know anything about Chinese, but if you are having problems with Ë, something like 見 would probably not work either. How about the file names?
It could also be pretty problematic handling Chinese in the same database (or table), since there are some problems between western languages already.
For example mysql's default collation for utf8 does think that a=å (å is a letter in Finnish and Swedish), so we have to use utf8swedish collation instead of the general one, which will probably not work too well with other languages.
Yet another thing to consider is the problem with URLs, if you have chinese characters in the video filenames.

**Comment by ErwinKomen** [comment]
Wouldn't Unicode problems be solved if the .pk instead of the idgloss is used in addressing glosses?
Database-internally it should work.
Only communication with the external world would still need to use the idgloss.

Verzonden met mijn Windows Phone
________________________________
Van: Henri<mailto:notifications@github.com>
Verzonden: ‎16-‎8-‎2016 18:02
Aan: Signbank/NGT-signbank<mailto:NGT-signbank@noreply.github.com>
Onderwerp: Re: [Signbank/NGT-signbank] Special character in gloss leads to errors (#194)


If you are planning to have Chinese glosses, this issue should be solved thoroughly. I don't know anything about Chinese, but if you are having problems with Ë, something like 見 would probably not work either. How about the file names?
It could also be pretty problematic handling Chinese in the same database (or table), since there are some problems between western languages already.
For example mysql's default collation for utf8 does think that a=å (å is a letter in Finnish and Swedish), so we have to use utf8swedish collation instead of the general one, which will probably not work too well with other languages.
Yet another thing to consider is the problem with URLs, if you have chinese characters in the video filenames.

**Comment by Woseseltops** [comment]
> Wouldn't Unicode problems be solved if the .pk instead of the idgloss is used in addressing glosses?

Correct, @ErwinKomen , that part of the problem was already solved (see my second post). 

The only problem I still had left was that I could not reach video files that had weird chars in their names via an url. I thought this was related to weird chars not being allowed in urls, as I write above, but today I accidently already solved the problem with this line:

https://github.com/Signbank/NGT-signbank/blob/dcf8664f1af8e59112712f1bcb8454d553c6da97/signbank/dictionary/models.py#L534

I believe there is still some percentage encoding going on under the hood... but you can now type weird characters in your browser, and it will correctly be linked to the video file containing weird characters in its name. 

> including Chinese glosses in a separate field, it'd be good to sort this out properly I think

The solution I now found seems to cover all Unicode characters, and apparently Unicode covers 74605 CJK (Chinese/Japanese/Korean) characters... which should be enough to start with :). To prove it, @ocrasborn, here is a video with a random Asian character in its name:

https://signbank.science.ru.nl//static/glossvideo/we/wesseltest6⾹-3533.mp4

and here is a testgloss with that same character:

https://signbank.science.ru.nl//dictionary/gloss/3533/

> It could also be pretty problematic handling Chinese in the same database 

I know your pain from experience, @henrinie , but so far everything seems to work out of the box, database-wise. I'm leaving things as they are for now :).

**Comment by ocrasborn** [comment]
Tried to add a gloss with just one chinese character (好）in ID Gloss and Annotation ID Gloss (Dutch), GOOD in Annotation ID Gloss (English). Resulting error:

UnicodeEncodeError at /dictionary/update/gloss/
'ascii' codec can't encode character u'\u597d' in position 0: ordinal not in range(128)

**Comment by ErwinKomen** [comment]
What about external communication: is it now still possible to 'predict' the URL of a signbank dictionary entry from its gloss in ELAN? Or is that not necessary, since the package-exchange provides ELAN with the correct match from idgloss to PK, hence to the URL?

If external predictability should still be required for some cases, then there is only one alternative solution I can come up with: use a concatenation (possibly interspersed with hyphens) of the unicode character numbers of the ID gloss letters in hexadecimal coding.

**Comment by Woseseltops** [comment]
> is it now still possible to 'predict' the URL of a signbank dictionary entry from its gloss in ELAN

It never was, I think: the url of an entry was and is [host]/dictionary/gloss/[pk]/ , so you will always need to know the pk of a gloss. If you do know the pk, you can also predict the video url: [host]/static/glossvideo/[first 2 letters of gloss]/[gloss]-[pk].mp4

The fact that the glosses, and thus the url of the video, can contain weird characters, should not be a problem: you can just type the weird character in your browser, and everything will work. The devtools show that under the hood, the browser transforms the weird chars with percent encoding; the last part of the video url above looks like wesseltest6%E2%BE%B9-3533.mp4 . (Maybe external software like ELAN may need to do this as well). However, Apache must already revert this to the actual weird chars, as the files stored on the server do not have any percent encoding in their names.

>  Or is that not necessary, since the package-exchange provides ELAN with the correct match from idgloss to PK, hence to the URL?

Our package functionality from issue #187 indeed provides external software like ELAN with a list of gloss-pk combinations.

**Comment by ocrasborn** [comment]
Addition to @Woseseltops 's first comment: it never used to be predictable, until we started using the pk as part of the ECV entry identifier in the ECV export. The ECV entries have the identifier gloss1234 (where gloss is a string not a variable). As we edit the ECV by making changes Signbank and then exporting to the ECV file on the server, we enforce that the pk is also used for new entries. e

**Comment by Woseseltops** [comment]
> Tried to add a gloss with just one chinese character (好）in ID Gloss and Annotation ID Gloss (Dutch), GOOD in Annotation ID Gloss (English). Resulting error

I could reproduce this... and it lead to an incredibly long journey through the SB code. The problem was (1) that there was an error message, but for the rest everything seemed to work as expected, and (2) the error message seem to refer to the page were the gloss was updated. After a long time I managed to pinpoint the problem to one specific line: if commented out the line that says `gloss.save()`, the error did not occur. Furthermore, saving a gloss only was a problem when one of the fields contained a weird char (newly added or previously). After an even longer time I noticed that the error message mentioned the reversion Django extension that we use (saves gloss history). This lead me to this page https://github.com/etianen/django-reversion/issues/111 , which explains that reversion expects the __unicode__ method to be set correclty. Once I did that, the problem was fixed.

Embarrassingly, @henrinie already suggested me to do exactly this in his very first comment to this thread, but I reverted the changed because it didn't seem to change anything back then :(.
